common-ts/src/utils/MultiplexWebSocket.ts

```diff
@@ -1,159 +1,189 @@
 import { Subject, catchError, Subscription } from 'rxjs';
 import WebSocket, { MessageEvent } from 'isomorphic-ws';
 
-export type WebSocketMessage = {
-	channel: string;
-	data: any;
-};
+type WebSocketMessage<T = Record<string, unknown>> = T;
 
-type WebSocketSubscriptionProps = {
+type WebSocketSubscriptionProps<T = Record<string, unknown>> = {
 	wsUrl: string;
 	subscriptionId: string;
 	subscribeMessage: string;
 	unsubscribeMessage: string;
 	onError: (err?: any) => void;
-	onMessage: (message: WebSocketMessage) => void;
-	messageFilter: (message: WebSocketMessage) => boolean;
-	errorMessageFilter: (message: WebSocketMessage) => boolean;
+	onMessage: (message: WebSocketMessage<T>) => void;
+	messageFilter: (message: WebSocketMessage<T>) => boolean;
+	errorMessageFilter: (message: WebSocketMessage<T>) => boolean;
 	onClose?: () => void;
 };
 
-type WebSocketSubscriptionState = WebSocketSubscriptionProps & {
-	hasSentSubscribeMessage?: boolean;
-	subjectSubscription?: Subscription;
-};
+type WebSocketSubscriptionState<T = Record<string, unknown>> =
+	WebSocketSubscriptionProps<T> & {
+		hasSentSubscribeMessage?: boolean;
+		subjectSubscription?: Subscription;
+	};
 
 type WebSocketUrl = string;
 type SubscriptionId = string;
 
+const DEFAULT_MAX_RECONNECT_ATTEMPTS = 5;
+const DEFAULT_MAX_RECONNECT_WINDOW_MS = 60 * 1000;
+
+class ReconnectionManager {
+	private reconnectAttempts: number = 0;
+	private lastReconnectWindow: number = Date.now();
+	private maxAttemptsCount: number;
+	private maxAttemptsWindowMs: number;
+
+	constructor(
+		maxAttemptsCount = DEFAULT_MAX_RECONNECT_ATTEMPTS,
+		maxAttemptsWindowMs = DEFAULT_MAX_RECONNECT_WINDOW_MS
+	) {
+		this.maxAttemptsCount = maxAttemptsCount;
+		this.maxAttemptsWindowMs = maxAttemptsWindowMs;
+	}
+
+	public attemptReconnection(wsUrl: string): {
+		shouldReconnect: boolean;
+		delay: number;
+	} {
+		const now = Date.now();
+
+		// Reset reconnect attempts if more than a minute has passed
+		if (now - this.lastReconnectWindow > this.maxAttemptsWindowMs) {
+			this.reconnectAttempts = 0;
+			this.lastReconnectWindow = now;
+		}
+
+		// Check if we've exceeded the maximum reconnect attempts
+		if (this.reconnectAttempts >= this.maxAttemptsCount) {
+			throw new Error(
```

Comment:
will it break the UI if the UI doesn't catch this error? is it better to just log the error?